common/scala/src/main/scala/whisk/core/containerpool/logging/ElasticSearchRestClient.scala

```diff
@@ -0,0 +1,211 @@
+      JsArray(JsObject(query.field -> JsObject("order" -> query.kind.toString.toJson)))
+  }
+
+  implicit object EsQuerySizeJsonFormat extends RootJsonFormat[EsQuerySize] {
+    def read(query: JsValue) = ???
+    def write(query: EsQuerySize) = JsNumber(query.size)
+  }
+
+  implicit object EsQueryAggsJsonFormat extends RootJsonFormat[EsQueryAggs] {
+    def read(query: JsValue) = ???
+    def write(query: EsQueryAggs) =
+      JsObject(query.aggField -> JsObject(query.agg.toString -> JsObject("field" -> query.field.toJson)))
+  }
+
+  implicit object EsQueryAllJsonFormat extends RootJsonFormat[EsQueryAll] {
+    def read(query: JsValue) = ???
+    def write(query: EsQueryAll) = JsObject("match_all" -> JsObject())
+  }
+
+  implicit object EsQueryMethod extends RootJsonFormat[EsQueryMethod] {
+    def read(query: JsValue) = ???
+    def write(method: EsQueryMethod) = method match {
+      case queryTerm: EsQueryTerm     => queryTerm.toJson
+      case queryString: EsQueryString => queryString.toJson
+      case queryMatch: EsQueryMatch   => queryMatch.toJson
+      case queryMust: EsQueryMust     => queryMust.toJson
+      case queryAll: EsQueryAll       => queryAll.toJson
+    }
+  }
+
+  implicit val esQueryFormat = jsonFormat4(EsQuery.apply)
+  implicit val esSearchHitFormat = jsonFormat(EsSearchHit.apply _, "_source")
+  implicit val esSearchHitsFormat = jsonFormat1(EsSearchHits.apply)
+  implicit val esSearchResultFormat = jsonFormat1(EsSearchResult.apply)
+}
+
+class ElasticSearchRestClient(
+  protocol: String,
+  host: String,
+  port: Int,
+  httpFlow: Option[Flow[(HttpRequest, Promise[HttpResponse]), (Try[HttpResponse], Promise[HttpResponse]), Any]] = None)(
+  implicit system: ActorSystem)
+    extends PoolingRestClient(protocol, host, port, 16 * 1024, httpFlow) {
+
+  import ElasticSearchJsonProtocol._
+
+  private val baseHeaders: List[HttpHeader] = List(Accept(MediaTypes.`application/json`))
+
+  /**
+   * Method to perform an ElasticSearch query that expects JSON response. If a payload is provided, a POST request will
+   * be performed. Otherwise, a GET request will be performed.
+   *
+   * @param uri     relative path to be used in the request to Elasticsearch
+   * @param headers list of HTTP request headers
+   * @param payload Optional JSON to be sent in the request
+   * @return Future with either the response with type T or the status code of the request, if the request is unsuccessful
+   */
+  def query[T: JsonFormat](uri: Uri,
+                           headers: List[HttpHeader] = List.empty,
+                           payload: Option[EsQuery] = None): Future[Either[StatusCode, T]] = {
+    val queryRes = payload match {
+      case Some(payload) =>
+        requestJson[JsObject](mkJsonRequest(HttpMethods.POST, uri, payload.toJson.asJsObject, baseHeaders ++ headers))
+
+      case None =>
+        requestJson[JsObject](mkRequest(HttpMethods.GET, uri, baseHeaders ++ headers))
+    }
```

Comment:
Can't `map` on a `match` result directly?

Reply:
Shouldn't it be `requestJson[T]` now? Should make belows code unneeded.

Reply:
For which cases is the `None` case even used? Is that used in any production code?

Reply:
Tried that, but got compile errors. Will see if I can get a workaround going.
```
/Users/dubee/bluewhisk/open/common/scala/src/main/scala/whisk/core/containerpool/logging/ElasticSearchRestClient.scala:184: Cannot find RootJsonReader or RootJsonFormat type class for T
        requestJson[T](mkJsonRequest(HttpMethods.POST, uri, payload.toJson.asJsObject, baseHeaders ++ headers))
                      ^
/Users/dubee/bluewhisk/open/common/scala/src/main/scala/whisk/core/containerpool/logging/ElasticSearchRestClient.scala:187: Cannot find RootJsonReader or RootJsonFormat type class for T
        requestJson[T](mkRequest(HttpMethods.GET, uri, baseHeaders ++ headers))
                      ^
```